Server/BranchingStrategy/ConflictLocator.cs

```diff
@@ -0,0 +1,88 @@
+﻿using PrincipleStudios.ScaledGitApp.Git;
+using PrincipleStudios.ScaledGitApp.Git.ToolsCommands;
+
+namespace PrincipleStudios.ScaledGitApp.BranchingStrategy;
+
+public class ConflictLocator(IGitToolsCommandInvoker gitToolsPowerShell, IGitConfigurationService gitConfiguration) : IConflictLocator
+{
+	class ConflictLocatorContext
+	{
+		public Dictionary<BranchPair, IReadOnlyList<IdentifiedConflict>> Results { get; } = [];
+		public required IReadOnlyDictionary<string, UpstreamBranchConfiguration> AllUpstreams { get; init; }
+
+		internal IReadOnlyList<string> GetImmediateUpstream(string branch)
+		{
+			return AllUpstreams.TryGetValue(branch, out var config) ? config.UpstreamBranchNames : Array.Empty<string>();
+		}
+	}
+
+	public async Task<IReadOnlyList<IdentifiedConflict>> FindConflictsWithin(IReadOnlyList<string> branches)
+	{
+		var results = new List<IdentifiedConflict>();
+		var context = await ConstructContext();
+		for (var i = 0; i < branches.Count - 1; i++)
+		{
+			results.AddRange(await FindConflictsBetween([branches[i]], branches.Skip(i + 1).ToArray(), context));
+		}
+		return results.ToArray();
+	}
+
+	public async Task<IReadOnlyList<IdentifiedConflict>> FindConflictsBetween(IReadOnlyList<string> leftBranches, IReadOnlyList<string> rightBranches)
+	{
+		var context = await ConstructContext();
+		return await FindConflictsBetween(leftBranches, rightBranches, context);
+	}
+
+	private async Task<IReadOnlyList<IdentifiedConflict>> FindConflictsBetween(IReadOnlyList<string> leftBranches, IReadOnlyList<string> rightBranches, ConflictLocatorContext context)
+	{
+		var leftCommonUpstream = context.AllUpstreams.GetCommonUpstream(leftBranches);
+		var rightCommonUpstream = context.AllUpstreams.GetCommonUpstream(rightBranches);
+
+		var result = new List<IdentifiedConflict>();
+		foreach (var left in leftBranches.Except(rightCommonUpstream))
+			foreach (var right in rightBranches.Except(leftCommonUpstream))
+			{
+				var additionalConflicts = await FindConflictsBetween(new BranchPair(left, right), context);
+				result.AddRange(additionalConflicts.Except(result).ToArray());
+			}
+		return result;
+	}
+
+	private async Task<IEnumerable<IdentifiedConflict>> FindConflictsBetween(BranchPair branchPair, ConflictLocatorContext context)
+	{
+		if (context.Results.TryGetValue(branchPair, out var cachedResult)) return cachedResult;
```

Comment:
How is this invalidated? Say I made a minor change to branchPair.RightBranch in order to resolve the conflict instead of creating an integration branch?

Reply:
Make a second request: this cache comes from the context of the one request, not a long-running in-memory cache.